src/lsp/lp_doc.ml

```diff
@@ -12,6 +12,106 @@
 
 open Core
 module LSP = Lsp_base
+(*
+exception NoPosition of string *)
+
+module type RangeMap = sig
+
+  type point
+  type interval
+  type 'a t
+
+  val create_point : int -> int -> point
+  val l : point -> int
+  val c : point -> int
+
+  val interval_start : interval -> point
+  val interval_end : interval -> point
+
+  val create_interval : point -> point -> interval
+
+  type cmp = Before | In | After
+
+  val in_range : point -> interval -> cmp
+
+  val add : interval -> 'a -> 'a t -> 'a t
+  val find : point -> 'a t -> 'a option * interval
+
+  val empty : 'a t
+
+  val rangemap_to_string : ('a * string) t -> string
+end
+
+module M : RangeMap = struct
+
+  type point = { l : int; c : int }
+
+  type interval = point * point
+
+  let create_point l c = { l = l; c = c}
+
+  let l pt = pt.l
+
+  let c pt = pt .c
+
+  let interval_start (x, _) = x
+
+  let interval_end (_, y) = y
+
+  let create_interval x y =
+    assert ( (l x < l y) || (l x = l y && c x <= c y) );
+    (x, y)
+
+  type cmp = Before | In | After
+
+  let in_range pos (pos1, pos2) =
+
+    if pos.l < pos1.l || (pos.l = pos1.l && pos.c < pos1.c)
+    then Before
+
+    else if pos.l > pos2.l || (pos.l = pos2.l && pos.c > pos2.c)
+    then After
+
+    else In
+
+  type 'a t = (interval * 'a) list
+
+  let rec find (cursor_pos : point) (l: 'a t) = match l with
+    |(interval, elt)::t ->
+      if in_range cursor_pos interval = In
+      then (Some elt, interval)
+
+      else find cursor_pos t
+    |_ -> (None, (create_point 0 0, create_point 0 0))
+
+  let add (k:interval) (elt:'a) (l :'a t) = (k, elt)::l
+
+  let empty = []
+
+  let rec rangemap_to_string (map :  (_ * string) t) = match map with
+    | [] -> "End"
+    | h::t -> let ( ({l = sl; c = sc}, {l = fl; c= fc}), (_ , str) ) = h in
+    (string_of_int sl)^","^(string_of_int sc)^"|"^(string_of_int fl)^","
+    ^(string_of_int fc)^":"^str^"\n"^(rangemap_to_string t)
+end
+
+let interval_of_pos (p : Pos.pos) = let open M in
+
+  let data = Lazy.force p in
+  let start : point = create_point data.start_line data.start_col in
+  let finish : point = create_point data.end_line data.end_col in
+
+  create_interval start finish
+
+let interval_of_popt (p : Pos.popt) = match p with
+  | None -> failwith "interval_of_popt : no position given for the token"
```

Comment:
invalid_arg instead?

Reply:
Better return and option type and have the clients handle `None`

Reply:
This function is not destined to the user but to a tricky fold_left so I don't know how to handle an option here.

Reply:
Actually the function is now used in the implementation of `RangeMap`, passed down to the stdlib's `Map.Make` fucntor. That seems quite tricky.